src/LanguageServer/Protocol/Handler/ServerLifetime/LspServiceLifeCycleManager.cs

```diff
@@ -14,14 +19,45 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler.ServerLifetime;
 internal sealed class LspServiceLifeCycleManager : ILifeCycleManager, ILspService
 {
     private readonly IClientLanguageServerManager _clientLanguageServerManager;
+    private readonly LspWorkspaceRegistrationService _lspWorkspaceRegistrationService;
 
-    public LspServiceLifeCycleManager(IClientLanguageServerManager clientLanguageServerManager)
+    [ExportCSharpVisualBasicLspServiceFactory(typeof(LspServiceLifeCycleManager)), Shared]
+    [method: ImportingConstructor]
+    [method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    internal class LspLifeCycleManagerFactory(LspWorkspaceRegistrationService lspWorkspaceRegistrationService) : ILspServiceFactory
+    {
+        public ILspService CreateILspService(LspServices lspServices, WellKnownLspServerKinds serverKind)
+        {
+            var clientLanguageServerManager = lspServices.GetRequiredService<IClientLanguageServerManager>();
+            return new LspServiceLifeCycleManager(clientLanguageServerManager, lspWorkspaceRegistrationService);
+        }
+    }
+
+    private LspServiceLifeCycleManager(IClientLanguageServerManager clientLanguageServerManager, LspWorkspaceRegistrationService lspWorkspaceRegistrationService)
     {
         _clientLanguageServerManager = clientLanguageServerManager;
+        _lspWorkspaceRegistrationService = lspWorkspaceRegistrationService;
     }
 
     public async Task ShutdownAsync(string message = "Shutting down")
     {
+        var hostWorkspace = _lspWorkspaceRegistrationService.GetAllRegistrations().SingleOrDefault(w => w.Kind == WorkspaceKind.Host);
+        if (hostWorkspace is not null)
+        {
+            var client = await RemoteHostClient.TryGetClientAsync(hostWorkspace, CancellationToken.None).ConfigureAwait(false);
+            if (client is not null)
+            {
+                await client.TryInvokeAsync<IRemoteExtensionMessageHandlerService>(
+                    (service, cancellationToken) => service.ResetAsync(cancellationToken),
+                    CancellationToken.None).ConfigureAwait(false);
+            }
+            else
+            {
+                var service = hostWorkspace.Services.GetRequiredService<IExtensionMessageHandlerService>();
```

Comment:
throw.

Reply:
or silently do nothing, or just log.   (same with all the other codepaths taht try to do something if we don't have a RHC.

Reply:
same as previous - this needs to work in VSCode which has no OOP